### PR TITLE
Add useKeyPress hook for subscribing to keyboard events

### PR DESCRIPTION
**Pull request by apomelitos**

### CODE REVIEW

File: src/components/Game.tsx [hunk 1]
```diff
@@ -26,6 +27,16 @@ const initialState: TileMeta[] = [
...
 
+  const moveLeft = () => console.log('moveLeft');
+  const moveRight = () => console.log('moveRight');
+  const moveUp = () => console.log('moveUp');
+  const moveDown = () => console.log('moveDown');
+
+  useKeyPress('ArrowLeft', moveLeft);
+  useKeyPress('ArrowRight', moveRight);
+  useKeyPress('ArrowUp', moveUp);
+  useKeyPress('ArrowDown', moveDown);
```
Comment (TheForsakenSpirit) [inline src/components/Game.tsx]: hide to one hook

**Review by TheForsakenSpirit — CHANGES_REQUESTED** [review]

File: src/hooks/useHandleButtons.ts [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+        emptyCellsPositions.push([rowIdx, colIdx]);
+      }
+    }
+  }
+
+  const position = emptyCellsPositions[Math.floor(Math.random() * emptyCellsPositions.length)];
+  console.log(emptyCellsPositions);
+
+  return {
+    id: Date.now(),
```
Comment (mrlika) [inline src/hooks/useHandleButtons.ts]: `Date.now()` depends on OS time. It means that I may break the game by changing OS time. Use `perfomance.now()` instead

File: src/hooks/useHandleButtons.ts [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+    for (let colIdx = 0; colIdx < SIZE; colIdx++) {
+      console.log(`Iteration: ${rowIdx * matrix.length + colIdx}`);
+      if (!isTile(matrix[rowIdx][colIdx])) {
+        emptyCellsPositions.push([rowIdx, colIdx]);
+      }
+    }
+  }
+
+  const position = emptyCellsPositions[Math.floor(Math.random() * emptyCellsPositions.length)];
+  console.log(emptyCellsPositions);
```
Comment (mrlika) [inline src/hooks/useHandleButtons.ts]: Don't push `console` statements into production

File: src/hooks/useHandleButtons.ts [hunk 4]
```diff
@@ -0,0 +1,140 @@
+import React, { SetStateAction, useEffect } from 'react';
+import { TileMeta } from '../types';
+
+const SIZE = 4;
+const getTilesMatrix = (tiles: TileMeta[]): TileMeta[][] => {
+  const matrix: TileMeta[][] = [];
+
+  for (let row = 0; row < SIZE; row++) {
+    matrix[row] = [];
+  }
```
Comment (TheForsakenSpirit) [inline src/hooks/useHandleButtons.ts]: Try to do this in one line with `new Array`, `fill` and `map`

File: src/hooks/useHandleButtons.ts [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+  const getNextPositionInRow = (pointer: number, idx: number): [number, number] => [idx, pointer];
+
+  let getTile = getTileFromRow;
+  let getNextPosition = getNextPositionInRow;
+  let pointerStart = 0;
+  let firstIdx = 0;
+  let changesCount = 0;
+
+  if (direction === 'UP' || direction === 'DOWN') {
+    getTile = getTileFromCol;
```
Comment (TheForsakenSpirit) [inline src/hooks/useHandleButtons.ts]: Avoid switching between function in one variable. Specialty mixing input data.

**Review by TheForsakenSpirit — CHANGES_REQUESTED** [review]

File: src/hooks/useHandleButtons.ts [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+        const position = getNextPosition(pointer, firstIdx);
+        newState.push({ ...tile, position, value });
+        pointer += pointerStep;
+
+        if (position[0] !== tile.position[0] || position[1] !== tile.position[1]) {
+          changesCount++;
+        }
+      }
+    }
+  }
```
Comment (mrlika) [inline src/hooks/useHandleButtons.ts]: The algorithm doesn't look correct. We will figure out this next week.

**Review by mrlika — CHANGES_REQUESTED** [review]

File: src/hooks/useHandleButtons.ts [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+    (direction: Direction) => {
+      let movedState: TileMeta[] = [];
+      let mergePairs: [TileMeta, TileMeta][] = [];
+      let changesCount = 0;
+      const options: Partial<MoveOptions> = {};
+
+      const getTileFromCol = (matrix: TileMeta[][], row: number, col: number): TileMeta => matrix[row][col];
+      const getTileFromRow = (matrix: TileMeta[][], col: number, row: number): TileMeta => matrix[row][col];
+      const getNextPositionInCol = (pointer: number, idx: number): Position => [pointer, idx];
+      const getNextPositionInRow = (pointer: number, idx: number): Position => [idx, pointer];
```
Comment (TheForsakenSpirit) [inline src/hooks/useHandleButtons.ts]: use one method to access.

File: src/hooks/useHandleButtons.ts [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+
+      if (direction === 'UP' || direction === 'DOWN') {
+        options.getTile = getTileFromCol;
+        options.getNextPosition = getNextPositionInCol;
+      } else {
+        options.getTile = getTileFromRow;
+        options.getNextPosition = getNextPositionInRow;
```
Comment (TheForsakenSpirit) [inline src/hooks/useHandleButtons.ts]: if you want conditional inverting access, create proxy-method that will conditionally invert column and row

**Review by TheForsakenSpirit — CHANGES_REQUESTED** [review]

**Review by TheForsakenSpirit — APPROVED** [review]

**Review by TheForsakenSpirit — APPROVED** [review]

**Review by mrlika — APPROVED** [review]

**Pull request merged**